trulens_eval/trulens_eval/feedback/benchmark_frameworks/tru_benchmark_experiment.py

```diff
@@ -0,0 +1,139 @@
+
+    def context_relevance_to_score(input, output, temperature=0):
+        return cortex.context_relevance(question=input, context=output, temperature=temperature)
+
+    benchmark_params = BenchmarkParams(temperature=0.5)
+    benchmark_experiment = TruBenchmarkExperiment(ground_truth, context_relevance_to_score, benchmark_params)
+
+    true_labels = benchmark_experiment.load_true_labels()
+    mae_agg_func = GroundTruthAggregator(true_labels=true_labels).mae
+    """
+
+    def __init__(
+        self,
+        ground_truth: Union[List, Callable, FunctionOrMethod],
+        trace_to_score_fn: Callable,
+        agg_funcs: List[AggCallable],
+        benchmark_params: BenchmarkParams,
+    ):
+        # TODO: instance type check of ground_truth argument + handle groundtruth_impl
+        self.ground_truth = ground_truth
+        self.trace_to_score_fn = trace_to_score_fn
+        self.benchmark_params = benchmark_params
+
+        self.f_benchmark_metrics: List[mod_feedback.Feedback] = [
+            Feedback(
+                lambda x: x,
+                name=f"metric_{agg_func.__name__}",
+            )
+            .on(Select.RecordCalls.run_score_generation_on_single_row.rets)
+            .aggregate(agg_func)
+            for agg_func in agg_funcs
+        ]
+
+    def load_beir_dataset(self, *args, **kwargs):
+        pass  # TODO
+
+    @instrument
+    def run_score_generation_on_single_row(
+        self, row, trace_to_score_fn: Callable
+    ) -> Union[float, Tuple[float, float]]:
+        """Generate a score with the trace_to_score_fn
+
+        Returns:
+            Union[float, Tuple[float, Dict[str, float]]]: feedback score (with metadata) after running the benchmark on a single entry in ground truth data
+        """
+
+        benchmark_params_dict: dict = self.benchmark_params.model_dump()
+        ret = trace_to_score_fn(
+            row["query"], row["response"], benchmark_params_dict
+        )
+
+        # TODO: better define the shape of arguments of trace_to_score_fn
+
+        if not isinstance(ret, tuple) and not isinstance(ret, float):
+            raise ValueError(
+                f"Output must be a float or a tuple, got {type(ret)}"
+            )
+
+        if isinstance(ret, tuple) and isinstance(ret[1], dict):
+            ret = (
+                ret[0],
+                list(ret[1].values())[-1],
+            )  # this is the case when a feedback function returns a tuple with a score and metadata like (0.5, {"confidence_score": 0.8})
+        return ret
+
+    @instrument
+    def collect_feedback_scores(
+        self,
+    ) -> Union[
+        List[float], List[Tuple[float]], Tuple[List[float], List[float]]
+    ]:
+        """Collect the list of generated feedback scores as input to the benchmark aggregation functions
+
+        Returns:
+            List[float]: feedback scores after running the benchmark on all entries in ground truth data
+        """
+        scores = []
```

Comment:
This should be a queue to account for concurrency?

Reply:
nvm, seems like the collection happens on the main thread. Just confirm if this is the case.

Reply:
yes it's the case with the threadpool logic below